packages/libs/client-utils/src/core/estimateGasPrice/utils.ts

```diff
@@ -0,0 +1,155 @@
+import { base64UrlDecode } from '@kadena/cryptography-utils';
+import { PactNumber } from '@kadena/pactjs';
+
+export const MINIMUM_GAS_PRICE = 1e-8;
+
+export const MAX_BLOCK_CAPACITY = 150000;
+
+export const median = (sortedArray: number[]): number => {
+  if (sortedArray.length === 0) {
+    return 0;
+  }
+  if (sortedArray.length % 2 === 1) {
+    return sortedArray[Math.floor(sortedArray.length / 2)];
+  } else {
+    return new PactNumber(sortedArray[sortedArray.length / 2 - 1])
+      .plus(sortedArray[sortedArray.length / 2])
+      .dividedBy(2)
+      .toNumber();
+  }
+};
+
+export const average = (arr: number[]): number => {
+  if (arr.length === 0) {
+    return 0;
+  }
+  return arr
+    .reduce((acc, gasFee) => acc.plus(gasFee), new PactNumber(0))
+    .dividedBy(arr.length)
+    .toNumber();
+};
+
+/**
+ * Interface representing the subset of structure of the block data returned from the Chainweb API
+ * this is not the full structure of the block data, but only the relevant parts for calculating gas fees.
+ */
+export interface IBlockChainData {
+  items: Array<{
+    header: {
+      height: number;
+    };
+    payloadWithOutputs: { transactions: Array<[string, string]> };
+  }>;
+}
+
+/**
+ * Interface representing the gas information for a block.
+ * This includes the total gas consumed, total gas limit, total gas paid, transaction count,
+ * block gas used percentage, and gas price statistics (min, max, avg, median).
+ * @public
+ */
+export interface IBlockGasInformation {
+  blockHeight: number;
+  totalGasConsumed: number; // sum of all gas consumed in the block transactions extracted from gas field of tx result
+  totalGasLimit: number; // sum of all gas limits in the block transactions extracted from gasLimit field of tx command
+  totalGasPaid: number; // sum of all gasPrice * gasLimit in the block transactions extracted from meta field of tx command
+  txCount: number; // number of transactions in the block
+  blockGasUsedPercent: number; // percentage of gas used in the block (totalGasConsumed / MAX_BLOCK_CAPACITY)
+  gasPriceStats: {
+    min: number; // minimum gas price in the block transactions
+    max: number; // maximum gas price in the block transactions
+    avg: number; // average gas price in the block transactions (does not consider usage)
+    median: number; // median gas price in the block transactions
+  };
+}
+
+/**
+ * Calculates gas information for a given set of blocks.
+ * @param blocks - An object containing block information.
+ * @returns An array of gas information for each block.
+ */
+export function calculateGasInformation(
```

Comment:
I'd add a try catch to catch the various potential errors from base64decoding and json parsing and provide more helpful errors.